pkg/smartcontract/context/context.go

```diff
@@ -0,0 +1,236 @@
+func (c *ParameterContext) AddSignature(ctr *wallet.Contract, pub *keys.PublicKey, sig []byte) error {
+	item := c.getItemForContract(ctr)
+	if vm.IsMultiSigContract(ctr.Script) {
+		if item.GetSignature(pub) != nil {
+			return errors.New("signature is already added")
+		}
+		pubs, err := parseScript(ctr.Script)
+		if err != nil {
+			return err
+		}
+		pubBytes := pub.Bytes()
+		var contained bool
+		for i := range pubs {
+			if bytes.Equal(pubBytes, pubs[i]) {
+				contained = true
+				break
+			}
+		}
+		if !contained {
+			return errors.New("public key is not present in script")
+		}
+		item.AddSignature(pub, sig)
+		if len(item.Signatures) == len(ctr.Parameters) {
+			indexMap := map[string]int{}
+			for i := range pubs {
+				indexMap[hex.EncodeToString(pubs[i])] = i
+			}
+			sigs := make([]sigWithIndex, 0, len(item.Signatures))
+			for pub, sig := range item.Signatures {
+				sigs = append(sigs, sigWithIndex{index: indexMap[pub], sig: sig})
+			}
+			sort.Slice(sigs, func(i, j int) bool {
+				return sigs[i].index < sigs[j].index
+			})
+			for i := range sigs {
+				item.Parameters[i] = smartcontract.Parameter{
+					Type:  smartcontract.SignatureType,
+					Value: sigs[i].sig,
+				}
+			}
+		}
+		return nil
+	}
+
+	index := -1
+	for i := range ctr.Parameters {
+		if ctr.Parameters[i].Type == smartcontract.SignatureType {
+			if index >= 0 {
+				return errors.New("multiple signature parameters in non-multisig contract")
+			}
+			index = i
+		}
+	}
+	if index == -1 {
+		return errors.New("missing signature parameter")
+	}
+	item.Parameters[index].Value = sig
+	return nil
+}
+
+func (c *ParameterContext) getItemForContract(ctr *wallet.Contract) *Item {
+	h := ctr.ScriptHash()
+	if item, ok := c.Items[h]; ok {
+		return item
+	}
+	params := make([]smartcontract.Parameter, len(ctr.Parameters))
+	for i := range params {
+		params[i].Type = ctr.Parameters[i].Type
+	}
+	item := &Item{
+		Script:     h,
+		Parameters: params,
+		Signatures: make(map[string][]byte),
+	}
+	c.Items[h] = item
+	return item
+}
+
+func parseScript(s []byte) ([][]byte, error) {
+	pubs := make([][]byte, 0)
+	ls := len(s)
+	for i := 0; i < ls; {
+		switch s[i] {
+		case byte(opcode.PUSHBYTES1):
+			i += 2
+		case byte(opcode.PUSHBYTES2):
+			i += 3
+		default:
+			i++
+		}
+		for i < ls && s[i] == byte(opcode.PUSHBYTES33) {
+			i++
+			if i+33 > ls {
+				return nil, errors.New("invalid CHECKMULTISIG script")
+			}
+			pubs = append(pubs, s[i:i+33])
+			i += 33
+		}
+	}
+	return pubs, nil
+}
+
+// MarshalJSON implements json.Marshaler interface.
+func (c ParameterContext) MarshalJSON() ([]byte, error) {
+	bw := io.NewBufBinWriter()
+	c.Verifiable.EncodeBinary(bw.BinWriter)
+	if bw.Err != nil {
+		return nil, bw.Err
+	}
+	items := make(map[string]json.RawMessage, len(c.Items))
+	for u := range c.Items {
+		data, err := json.Marshal(c.Items[u])
+		if err != nil {
+			return nil, err
+		}
+		items["0x"+u.StringBE()] = data
```

Comment:
Are you sure about BE here?

Reply:
Oh, it should be LE

Reply:
Actually no, it's `BE` as in `(util.Uint160).MarshalJSON()`

Reply:
OK, so be it.